README.md
# imc_ros_bridge
![CI](https://github.com/smarc-project/imc_ros_bridge/workflows/CI/badge.svg?branch=noetic-devel) [![license](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

Minimal library for bridging ROS and IMC messages

## Building

A simple `catkin_make` should do.

## Launching

The command
```
roslaunch imc_ros_bridge bridge.launch server_addr:=127.0.0.1 server_port:=6002
```
will launch the `imc_bridge` node.

## ROS to IMC conversion and use on vehicles

See the [vehicles documents](https://github.com/smarc-project/imc_ros_bridge/blob/noetic-devel/docs/vehicles.md).

## Trying with Neptus

These instructions have been tested for Ubuntu 16.04, 18.04, 20.04. For 18.04 and 20.04, you
may have to comment a line in a jdk file to get it running (See Troubleshooting).

First, install dependencies: `sudo apt install openjdk-11-jdk`.
Second, make sure there are no other Javas in your system, you might have a `jre-default` installed by default. Use `apt remove ...` to remove it (See Troubleshooting for more details).

Clone and build Neptus at a known-working commit:
```
git clone https://github.com/LSTS/neptus.git
git checkout 38c7f41a9885c6b059f79b38861edb4b7b67511b
cd neptus
./gradlew
```
If successful, run it with `./neptus`.
As of Mar 2 2021, this will open up a map view and you should see the vehicles in this map.
(Neptus is being actively developed nowadays, so things might change, contact @KKalem if need be)


If you go to the systems list panel (top bar: view), you should also see the ROS auv displayed as a cyan panel.
This indicates that neptus can communicate with the auv. Any other color indicates som problem,
see [this link](https://www.lsts.pt/neptus/manual/trunk/elements.html#systems-list).

### For SMaRC Vehicles:
Delete the `neptus/vehicle-defs` folder and copy `imc_ros_bridge/vehicle-defs` into `neptus/` instead (You could also merge the two instead of replacing). This way, in Neptus, you will (only) see SAM and LOLO as options and Neptus will know the maneuvers they are capable of.
You will need to update these files manually if they ever change in the future.
(As of Mar 2 2021, the vehicle-files are not used because the 'main window' of Neptus no longer exists, this might change in the future)


### Troubleshooting
If Neptus does not seem to work, try these:
On Ubuntu 18.04 and 20.04, you might have `openjdk-11-jre-headless` and `default-jre` remove them with `apt remove ...` manually and make sure that `apt list --installed | grep jre` only shows one jre, it should look like this:
```
$ apt list --installed | grep jre

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

openjdk-8-jre-headless/focal-updates,focal-security,now 8u282-b08-0ubuntu1~20.04 amd64 [installed,automatic]
openjdk-8-jre/focal-updates,focal-security,now 8u282-b08-0ubuntu1~20.04 amd64 [installed,automatic]
```
This is a known-working setup as of Mar 2 2021.


Additionally, if you get a "No VTK Java Packages Found" message in Neptus's console AND things are not working, do this(might need to change the java version to the version you have):

`sudo vim /etc/java-8-openjdk/accessibility.properties`
Comment out the following line:
`assistive_technologies=org.GNOME.Accessibility.AtkWrapper`

If you get an error that involves `iced-tea` in the Neptus terminal window when you try to open a console, check that you have openjdk-8-jre. Not _just_ the headless version of it, Neptus needs the head.


## Existing conversions

### ros_to_imc

* `std_msgs::Empty` on topic `/heartbeat` -> `IMC::Heartbeat`
* `sensor_msgs::NavSatFix` on topic `/gps_fix` -> `IMC::GpsFix`
* `geometry_msgs::Pose` on topic `/goto_input` -> `IMC::Goto`
* `sensor_msgs::NavSatFix` on topic `/gps_nav_data` -> `IMC::GpsNavData`
* `sensor_msgs::NavSatFix` on topic `/estimated_state` -> `IMC::EstimatedState`
* `imc_ros_bridge::PlanControlState` on topic `plan_control_state` -> `IMC::PlanControlState`

### imc_to_ros

* `IMC::Goto` -> `geometry_msgs::Pose` on topic `/goto_waypoint`
* `IMC::Abort` -> `std_msgs::Empty` on topic `/abort`
* `IMC::Heartbeat` -> `std_msgs::Empty` on topic `/imc_heartbeat`
* `IMC::PlanDB` -> `std_msgs::String` on topic `/plan_db` (JSON)
* `IMC::Plancontrol` -> `imc_ros_bridge::PlanControl` on topic `/plan_control`

## Creating new conversions

For each conversion in either direction, you need to create a new library
that contains a specialized `convert` function for the types you want to convert.
You then need to add a `BridgeServer` to the node, see details below.

### ros_to_imc

Check out examples in the `include/imc_ros_bridge/ros_to_imc` and `src/ros_to_imc` folders
for more examples. All conversions specialize the `convert` function like this:

```cpp
namespace ros_to_imc {
template <>
bool convert(const sensor_msgs::NavSatFix& ros_msg, IMC::GpsFix& imc_msg)
{
    imc_msg.lat = ros_msg.latitude;
    imc_msg.lon = ros_msg.longitude;
    imc_msg.height = ros_msg.altitude;

    return true;
}
}
```

And they also add the bridge server to the `bridge_node` like this:
```cpp
ros_to_imc::BridgeServer<sensor_msgs::NavSatFix, IMC::GpsFix> gpsfix_server(ros_node, imc_handle, "/gps_fix");
```
And link the convert libary into `ros_to_imc_node` in the `CMakeLists.txt` file.

### imc_to_ros

Check out examples in the `include/imc_ros_bridge/imc_to_ros` and `src/imc_to_ros` folders
for more examples. All conversions specialize the `convert` function like this:

```cpp
namespace imc_to_ros {
template <>
bool convert(const IMC::Goto& imc_msg, geometry_msgs::Pose& ros_msg)
{
    ros_msg.position.x = imc_msg.lon;
    ros_msg.position.y = imc_msg.lat;
    ros_msg.position.z = imc_msg.z;
    return true;
}
}
```

And they also add the bridge server to the `bridge_node` like this:
```cpp
imc_to_ros::BridgeServer<IMC::Goto, geometry_msgs::Pose> goto_server(imc_handle, ros_node, "/goto_waypoint");
```
And link the convert libary into `imc_to_ros_node` in the `CMakeLists.txt` file.

## Neptus-AUV Integration
For some vehicle named `$VEHICLE`.
In this repo, we have three vehicles available: `sam-auv`, `lolo-auv` and `imc_ros_bridge`.


### IMPORTANT NOTE

The bridge should be running on the same physical device as Neptus. This is a current limitation due to IP address shenanigans that is planned to be fixed eventually. Both for physical and simulated vehicles, Neptus and bridge must run on the same device.


### Adding a vehicle to Neptus

An arbitrary integer `$NUMBER` is used by Neptus as a convention to group together similar vehicles. 

Copy `neptus_vehicle_definitions/$VEHICLE_files/$NUMBER-$VEHICLE.nvcl` into `.../neptus/vehicle-defs/` and `neptus_vehicle_definitions/$VEHICLE_files/$VEHICLE` folder into `.../neptus/vehicle_files/`.
This will add `$VEHICLE` to the list of vehicles available in Neptus' list, with `$VEHICLE`'s visuals. 

Do the same thing with `imc_ros_bridge` to get a generic vehicle. The `imc_ros_bridge` fake vehicle has id 4.

When starting the `imc_ros_brdige` node, be careful to set the `imc_id` parameter to the correct one, otherwise the bridge will not receive data from the respective console window. Do not forget to select the correct vehicle in the console window (top right, drop down menu) either. 

The `.nvcl` file contains the id of the vehicle you are adding. You can find the `imc-id` field under `<communication-means> -> <comm-mean> -> <protocols-args> -> <imc>`. This field should be changed to whatever imc id you might want to use. It also must match the `imc_id` parameter used by the bridge node.

For advanced uses of the `.nvcl` file, consult Neptus documentation.

### Moving a vehicle around

Using the console (opened by going to Vehicles -> `$VEHICLE` -> click console button) generate a plan (In the console window: Tools -> Generate plan or use the plan tool on the left). Select the plan on the right section of the console and click the blue arrow (upload button) towards the top. This will create and send out a JSON formatted plan to the ros topic `/plan_db`. Parse this JSON on the vehicle, and control `$VEHICLE`.

In order to see the updated pose of $VEHICLE in the Neptus console, publish to the ros topic `/estimated_state`. Currently only lat, lon, altitude are used. The update rate on the Neptus console can be about once every 1-2 seconds, be patient.

See [this tutorial](https://github.com/smarc-project/smarc_scenarios/tree/master/bts_tutorial) for a demo that uses this bridge as part of a full planning -> execution pipeline in simulation. 

### Emergency

From the Neptus console, the big red ABORT button can be used to send an empty message to the ros topic `/abort`. Probably a good idea to subscribe to this topic. Disregard the errors that pop out in the Neptus console due to missing acoustic transponders and such.








CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package imc_ros_bridge
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2022-03-30)
------------------
* Update README.md
* Add files via upload
* Create vehicles.md
* Update release.yml
* Merge pull request `#26 <https://github.com/smarc-project/imc_ros_bridge/issues/26>`_ from KKalem/noetic-devel
  Added CoverArea maneuver
* Added CoverArea maneuver
* Merge pull request `#25 <https://github.com/smarc-project/imc_ros_bridge/issues/25>`_ from KKalem/noetic-devel
  Changed lolo's defaults to -30m and 200rpm for goto/sample maneuvers
* Changed lolo's defaults to -30m and 200rpm for goto/sample maneuvers
* Merge pull request `#24 <https://github.com/smarc-project/imc_ros_bridge/issues/24>`_ from KKalem/noetic-devel
  New "Sample" maneuver and imc_id semantics changes
* Added sample maneuver to lolo, removed unhandled maneuvers, added lolos picture
* Added handling for the SAMPLE maneuver, changed the imc_id imc_src semantics to match Neptus, started to ignore tcp_callback for planDB messages that were spammy, added new fields to Maneuver.msg for the SAMPLE maneuver
* Added SAMPLE maneuver to SAM, added SAM's picture to vehicle defs
* Merge pull request `#23 <https://github.com/smarc-project/imc_ros_bridge/issues/23>`_ from KKalem/noetic-devel
  Updated instructions and files for the new Neptus version.
* Removed unused files and renamed the folder for easier copying
* Updated README for the newer version of Neptus
* Update release.yml
* Merge pull request `#22 <https://github.com/smarc-project/imc_ros_bridge/issues/22>`_ from nilsbore/test_release
  Release building
* Create release.yml
* Added install of targets
* Merge pull request `#21 <https://github.com/smarc-project/imc_ros_bridge/issues/21>`_ from smarc-project/lab_test
  Put node in imc namespace
* Added install rules
* Put node in imc namespace
* Merge pull request `#20 <https://github.com/smarc-project/imc_ros_bridge/issues/20>`_ from smarc-project/nilsbore-patch-1
  Update main.yml
* Update main.yml
* Update README.md
* Update package.xml
* Update package.xml
* Create main.yml
* Merge pull request `#18 <https://github.com/smarc-project/imc_ros_bridge/issues/18>`_ from KKalem/master
  Slight improvement to the md5 situation.
* Changed md5 calcs to be 1-to-1 to java version. md5s still not matching with neptus, but this is a better start than before
* Corrected the md5 format. Removed some noise
* Merge pull request `#17 <https://github.com/smarc-project/imc_ros_bridge/issues/17>`_ from KKalem/master
  Added plandbinformation and plandbstate messages for better neptus in…
* Added plandbinformation and plandbstate messages for better neptus integration. Somehow the md5 we calculate here is different than the one Neptus calculates and compares, will fix later.
* Merge pull request `#16 <https://github.com/smarc-project/imc_ros_bridge/issues/16>`_ from KKalem/master
  Added ros to imc planDB message xlation so we can answer neptus
* Added ros to imc planDB message xlation so we can answer neptus's requests
* Merge pull request `#15 <https://github.com/smarc-project/imc_ros_bridge/issues/15>`_ from KKalem/master
  Made multiple bridges/vehicles possible.
* Added new imc_src argument to udp_link. imc messages have src_ent to separate different vehicles of the same type and id. imc_src is passed on to src_ent of every message
* Merge pull request `#14 <https://github.com/smarc-project/imc_ros_bridge/issues/14>`_ from KKalem/master
  Added new launch file, cleanup
* Added new launch file, bridge_sam which has sam-specific defaults and calls bridge.launch with those, cleanup
* Merge pull request `#13 <https://github.com/smarc-project/imc_ros_bridge/issues/13>`_ from KKalem/master
  Added proper planDB message. At least the GOTO maneuvers are in there now.
* Merge branch 'master' of https://github.com/smarc-project/imc_ros_bridge
* Added new PlanDB message _and all the billion sub-messages it needed\_
* Merge pull request `#10 <https://github.com/smarc-project/imc_ros_bridge/issues/10>`_ from NiklasRolleberg/master
  rosparam for ip of bridge and netpus + more conversions + compilation fix on jetson
* Merge pull request `#2 <https://github.com/smarc-project/imc_ros_bridge/issues/2>`_ from NiklasRolleberg/master
  merge niklas initial
* Added RemoteState conversion
* Merge branch 'master' into master
* Added parameter for chaging the name of the node and increased the precission of floats in missions to get better accuracy of waypoints
* Merge pull request `#12 <https://github.com/smarc-project/imc_ros_bridge/issues/12>`_ from KKalem/master
  Added namespacing to bridge.launch.
* Added the old Estimated state conversion again
* Merge branch 'lidingo_changes'
* Added namepsacing to bridge launch file. default to sam
* Merge pull request `#11 <https://github.com/smarc-project/imc_ros_bridge/issues/11>`_ from KKalem/master
  imc_id changes so that they are consistent. Also small README update.
* Merge branch 'master' into master
* Update README.md
* Merge branch 'master' of https://github.com/smarc-project/imc_ros_bridge
* Changed default imc_ids to be consistent
* Merge pull request `#9 <https://github.com/smarc-project/imc_ros_bridge/issues/9>`_ from KKalem/master
  Updated README with JRE installation clarification for ubuntu 18
* changed namespace
* removed logging
* compiling on jetson
* Added a few new conversions
* removed a line
* changed name of parameters related to ip and port
* Added important note to README
* added a simple way of logging and changed Estimatedstate converter
* Removed namespacing from the bridge in favor of the launch file namespacing method
* Update README with more instructions for Ubuntu 18
* Updated README with JRE installation clarification for ubuntu 18
* Added sonardata message and increased send buffer size to be able to send bigger messages
* changed to x=lat & y = lon in pose message to fit better with the NED frame

src/bridge_node.cpp
/* Copyright 2019 The SMaRC project (https://smarc.se/)
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <ros/ros.h>
#include <iostream>

#include <imc_ros_bridge/imc_ros_bridge_server.h>

#include <imc_ros_bridge/ros_to_imc/Heartbeat.h>
#include <imc_ros_bridge/ros_to_imc/GpsFix.h>
#include <imc_ros_bridge/ros_to_imc/Goto.h>
#include <imc_ros_bridge/ros_to_imc/GpsNavData.h>
#include <imc_ros_bridge/ros_to_imc/EstimatedState.h>
#include <imc_ros_bridge/ros_to_imc/PlanControlState.h>
#include <imc_ros_bridge/ros_to_imc/VehicleState.h>
#include <imc_ros_bridge/ros_to_imc/RemoteState.h>
#include <imc_ros_bridge/ros_to_imc/SonarData.h>
#include <imc_ros_bridge/ros_to_imc/DesiredHeading.h>
#include <imc_ros_bridge/ros_to_imc/DesiredHeadingRate.h>
#include <imc_ros_bridge/ros_to_imc/DesiredPitch.h>
#include <imc_ros_bridge/ros_to_imc/DesiredRoll.h>
#include <imc_ros_bridge/ros_to_imc/DesiredSpeed.h>
#include <imc_ros_bridge/ros_to_imc/DesiredZ.h>
#include <imc_ros_bridge/ros_to_imc/PlanDB.h>

#include <imc_ros_bridge/imc_to_ros/Goto.h>
#include <imc_ros_bridge/imc_to_ros/Heartbeat.h>
#include <imc_ros_bridge/imc_to_ros/Abort.h>
#include <imc_ros_bridge/imc_to_ros/PlanDB.h>
#include <imc_ros_bridge/imc_to_ros/PlanControl.h>


using namespace std;

int main(int argc, char** argv)
{
    ros::init(argc, argv, "imc_bridge");
    ros::NodeHandle ros_node;

    std::string neptus_addr;
    std::string bridge_tcp_addr;
    std::string bridge_tcp_port;
    std::string sys_name;
    int imc_id;
	int imc_src;
    ros::param::param<std::string>("~neptus_addr", neptus_addr, "127.0.0.1");
    ros::param::param<std::string>("~bridge_addr", bridge_tcp_addr, "127.0.0.1");
    ros::param::param<std::string>("~bridge_port", bridge_tcp_port, "6002");
    ros::param::param<std::string>("~system_name", sys_name, "imc_ros_bridge");
	// Neptus seems to have swapped these values for SOME REASON. so I swap here too...
	// tldr: imc_id is the vehicle TYPE and imc_src is the SPECIFIC VEHICLE
    ros::param::param<int>("~imc_id", imc_src, 30);
	ros::param::param<int>("~imc_src", imc_id, 5);

    IMCHandle imc_handle(bridge_tcp_addr, bridge_tcp_port, neptus_addr, sys_name, imc_id, imc_src);

    ros_to_imc::BridgeServer<std_msgs::Empty, IMC::Heartbeat> heartbeat_server(ros_node, imc_handle, "heartbeat");
    ros_to_imc::BridgeServer<sensor_msgs::NavSatFix, IMC::GpsFix> gpsfix_server(ros_node, imc_handle, "gps_fix");
    ros_to_imc::BridgeServer<geometry_msgs::Pose, IMC::Goto> goto_server_dummy(ros_node, imc_handle, "goto_input");
    ros_to_imc::BridgeServer<sensor_msgs::NavSatFix, IMC::GpsNavData> gpsnavdata_server(ros_node, imc_handle, "gps_nav_data");
    ros_to_imc::BridgeServer<imc_ros_bridge::EstimatedState, IMC::EstimatedState> estimatedstate_server(ros_node, imc_handle, "estimated_state");
    ros_to_imc::BridgeServer<imc_ros_bridge::PlanControlState, IMC::PlanControlState> plancontrolstate_server(ros_node, imc_handle, "plan_control_state");
    ros_to_imc::BridgeServer<imc_ros_bridge::VehicleState, IMC::VehicleState> vehiclestate_server(ros_node, imc_handle, "vehicle_state");
    ros_to_imc::BridgeServer<imc_ros_bridge::RemoteState, IMC::RemoteState> RemoteState_server(ros_node, imc_handle, "remote_state");
    ros_to_imc::BridgeServer<imc_ros_bridge::SonarData, IMC::SonarData> sonar_server(ros_node, imc_handle, "sonar_data");
    ros_to_imc::BridgeServer<imc_ros_bridge::DesiredHeading, IMC::DesiredHeading> DesiredHeading_server(ros_node, imc_handle, "desired_heading");
    ros_to_imc::BridgeServer<imc_ros_bridge::DesiredHeadingRate, IMC::DesiredHeadingRate> DesiredHeadingRate_server(ros_node, imc_handle, "desired_heading_rate");
    ros_to_imc::BridgeServer<imc_ros_bridge::DesiredPitch, IMC::DesiredPitch> DesiredPitch_server(ros_node, imc_handle, "desired_pitch");
    ros_to_imc::BridgeServer<imc_ros_bridge::DesiredRoll, IMC::DesiredRoll> DesiredRoll_server(ros_node, imc_handle, "desired_roll");
    ros_to_imc::BridgeServer<imc_ros_bridge::DesiredSpeed, IMC::DesiredSpeed> DesiredSpeed_server(ros_node, imc_handle, "desired_speed");
    ros_to_imc::BridgeServer<imc_ros_bridge::DesiredZ, IMC::DesiredZ> DesiredZ_server(ros_node, imc_handle, "desired_z");
    ros_to_imc::BridgeServer<imc_ros_bridge::PlanDB, IMC::PlanDB> to_imc_plandb_server(ros_node, imc_handle, "plan_db");
  
    // 450
    imc_to_ros::BridgeServer<IMC::Goto, geometry_msgs::Pose> goto_server(imc_handle, ros_node, "goto_waypoint");
    // 150
    imc_to_ros::BridgeServer<IMC::Heartbeat, std_msgs::Empty> imc_heartbeat_server(imc_handle, ros_node, "imc_heartbeat");
    // 550
    imc_to_ros::BridgeServer<IMC::Abort, std_msgs::Empty> abort_server(imc_handle, ros_node, "abort");
    // 556
    imc_to_ros::BridgeServer<IMC::PlanDB, imc_ros_bridge::PlanDB> plandb_server(imc_handle, ros_node, "plan_db");
    imc_to_ros::BridgeServer<IMC::PlanControl, imc_ros_bridge::PlanControl> plancontrol_server(imc_handle, ros_node, "plan_control");

    auto announce_callback = [&](const ros::TimerEvent&) { imc_handle.announce(); };
    auto heartbeat_callback = [&](const ros::TimerEvent&) { imc_handle.publish_heartbeat(); };

    ros::Timer announce_timer = ros_node.createTimer(ros::Duration(10.), announce_callback);
    ros::Timer heartbeat_timer = ros_node.createTimer(ros::Duration(1.), heartbeat_callback);

    ros::spin();

    return 0;
}
